overlord/devicestate/devicemgr.go

```diff
@@ -1425,6 +1425,73 @@ func (m *DeviceManager) ensurePostFactoryReset() error {
 	return os.Remove(factoryResetMarker)
 }
 
+func (m *DeviceManager) hasUserExpired(user *auth.UserState) bool {
+	// If the user has no expiration date, then Expiration should not
+	// be set, and contain the default value.
+	if user.Expiration.IsZero() {
+		return false
+	}
+
+	return user.Expiration.Before(time.Now())
+}
+
+var osutilDelUser = osutil.DelUser
+
+func (m *DeviceManager) removeExpiredUser(st *state.State, user *auth.UserState) error {
+	// Remove the user from the system first
+	if err := osutilDelUser(user.Username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return err
+	}
+
+	// And then we remove it from the auth state
+	_, err := auth.RemoveUserByUsername(st, user.Username)
```

Comment:
I would expect these two function calls to be swapped. If I'm reading the code right, in `daemon/api_users.go` we first call `osutilAddUser()` and later call `setupLocalUser()`, so it seems more logical if the delition happens in the reverse order. If there are reasons why it's shouldn't be done in that way, it would be nice to add a code comment with them.

Reply:
In `daemon/api_users.go` the deletion occurs in this order, but I can see your point. There is no explanation as to why the user is deleted in this order. My theory is that there might be some conditions under where the user cannot be removed/fails to be removed from the system, and if the user was *not* removed by the system, then it's crucial it not be removed from auth state.

Reply:
OK, fine with me then. But let's leave this conversation open so that Samuele can see it and weigh in if needed.

Reply:
Resolving this issue as this code has been removed